File: 2023/challenge_05/challenge_05.py
import os
import csv
import re

def read_file():
  path = os.path.dirname(__file__)
  with open(os.path.join(path, 'challenge_05.csv'), encoding='UTF-8') as file:
    filed_names = ['id', 'username', 'email', 'age', 'location']
    return list(csv.DictReader(file, fieldnames=filed_names))

def validate_id(user_id):
  return user_id and user_id.isalnum()

def validate_username(username):
  return username and username.isalnum()

def validate_email(email):
  return email and re.match(r'[^@]+@[^@]+\.[^@]+', email)

def validate_age(age):
  return not age or age.isdigit()

def validate_location(location):
  return not location or re.match(r'^[\w\s]+$', location)

def run():
  message = ''
  users = read_file()
  
  for user in users:
    if not (validate_id(user['id']) and \
        validate_username(user['username']) and \
        validate_email(user['email']) and \
        validate_age(user['age']) and \
        validate_location(user['location'])):
      message += user['username'][0]
    
  return message

if __name__ == "__main__":
  print(run())

File: 2023/challenge_05/challenge_05.js
const fs = require('node:fs/promises')
const path = require('node:path')

async function read_file() {
  const text = await fs.readFile(path.join(__dirname, 'challenge_05.csv'), 'utf8');
  return text.split('\n');
}

const VALIDATIONS = {
  0: id => id && /^[a-zA-Z0-9]+$/.test(id),
  1: username => username && /^[a-zA-Z0-9]+$/.test(username),
  2: email => email && /\w+@\w+\.\w+/.test(email),
  3: age => !age || /^[0-9]+$/.test(age),
  4: location => !location || /^[\w\s]+$/.test(location)
}

async function run() {
  const users = await read_file();
  let message = '';

  users.forEach(user => {
    const userInfo = user.trim().split(',');
    let wrongUser = false;
  
    for (let i = 0; i < userInfo.length; i++) {
      if (!VALIDATIONS[i](userInfo[i])) {
        wrongUser = true;
        break;
      }
    }
    
    if (wrongUser) message += userInfo[1][0];
  });

  return message
}

run().then(console.log)